keygen.go

```diff
@@ -120,119 +139,124 @@ func NewWithWrite(path, name string, passphrase []byte, keyType KeyType) (*SSHKe
 	if err != nil {
 		return nil, err
 	}
-	if !s.IsKeyPairExists() {
+	if !s.KeyPairExists() {
 		if err = s.WriteKeys(); err != nil {
 			return nil, err
 		}
 	}
 	return s, nil
 }
 
-// generateEd25519Keys creates a pair of EdD25519 keys for SSH auth.
-func (s *SSHKeyPair) generateEd25519Keys() error {
-	// Generate keys
-	pubKey, privateKey, err := ed25519.GenerateKey(rand.Reader)
+// PrivateKey returns the unencrypted private key in OPENSSH PEM format.
+func (s *SSHKeyPair) PrivateKey() []byte {
+	block, err := s.pemBlock()
 	if err != nil {
-		return err
+		return nil
 	}
+	return pem.EncodeToMemory(block)
+}
 
-	// Encode PEM
-	pemBlock := pem.EncodeToMemory(&pem.Block{
-		Type:  "OPENSSH PRIVATE KEY",
-		Bytes: edkey.MarshalED25519PrivateKey(privateKey),
-	})
-
+// PublicKey returns the SSH public key (RFC 4253). Ready to be used in an
+// OpenSSH authorized_keys file.
+func (s *SSHKeyPair) PublicKey() []byte {
+	var err error
+	var p ssh.PublicKey
 	// Prepare public key
-	publicKey, err := ssh.NewPublicKey(pubKey)
+	switch s.keyType {
+	case RSA:
+		p, err = ssh.NewPublicKey(s.rsaKey.Public())
+	case Ed25519:
+		p, err = ssh.NewPublicKey(s.ed25519Key.Public())
+	case ECDSA:
+		p, err = ssh.NewPublicKey(s.ecdsaKey.Public())
+	default:
+		return nil
+	}
 	if err != nil {
-		return err
+		return nil
 	}
+	// serialize public key
+	ak := ssh.MarshalAuthorizedKey(p)
+	return pubKeyWithMemo(ak)
+}
 
-	// serialize for public key file on disk
-	serializedPublicKey := ssh.MarshalAuthorizedKey(publicKey)
+func (s *SSHKeyPair) pemBlock() (block *pem.Block, err error) {
+	block = &pem.Block{}
+	switch s.keyType {
+	case RSA:
+		block.Type = "RSA PRIVATE KEY"
+		block.Bytes = x509.MarshalPKCS1PrivateKey(s.rsaKey)
+	case Ed25519:
+		block.Type = "OPENSSH PRIVATE KEY"
+		block.Bytes = edkey.MarshalED25519PrivateKey(*s.ed25519Key)
+	case ECDSA:
+		block.Type = "EC PRIVATE KEY"
+		block.Bytes, err = x509.MarshalECPrivateKey(s.ecdsaKey)
+	default:
+		return nil, fmt.Errorf("unsupported key type %s", s.keyType)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return block, nil
+}
 
-	s.PrivateKeyPEM = pemBlock
-	s.PublicKey = pubKeyWithMemo(serializedPublicKey)
-	return nil
+func (s *SSHKeyPair) encryptedPrivateKey(passphrase []byte) ([]byte, error) {
+	block, err := s.pemBlock()
+	if err != nil {
+		return nil, err
+	}
+	switch block.Type {
+	case "OPENSSH PRIVATE KEY":
+		return sshkeys.Marshal(s.ed25519Key, &sshkeys.MarshalOptions{Passphrase: passphrase})
+	default:
+		// encrypt private key with passphrase
+		block, err = x509.EncryptPEMBlock(rand.Reader, block.Type, block.Bytes, passphrase, x509.PEMCipherAES256)
```

Comment:
due to this being insecure, and the `sshkeys` also using it, and this PR would be a breaking change anyway... maybe its worth not supporting passwords at all?

I personally never use it anyway 🤔

thoughts?

Reply:
Good point, `x509.EncryptPEMBlock()` is deprecated and currently, there is no alternative. Although almost no one uses password-protected keys, it is a nice feature to have.

Maybe we could support parsing password-protected keys and errors on generating RSA or ECDSA protected keys. ED25519 is fine since it doesn't use `x509.EncryptPEMBlock()`. At least until Golang supports marshaling OpenSSH formats https://github.com/golang/go/issues/37132 and https://go-review.googlesource.com/c/crypto/+/218620/

Reply:
Hi was about to open an issue for passphrase support 😅! I would love to see passphrase support for ED25519 keys. At work, we are required to passphrase protect all out ssh-keys  

It would allow me to use keygen to streamline some internal tooling. 🙂

Reply:
Using https://github.com/caarlos0/sshmarshal until https://go-review.googlesource.com/c/crypto/+/218620/ becomes upstream